In [12]:
import numpy as np
import pandas as pd
from sklearn.linear_model import LogisticRegression

In [13]:
data = pd.read_csv(r"C:\Users\nico_\Desktop\fichiers_csv\traitement.csv", sep=";")

sex (binary): 1 if Male, 0 if Female

age (int): age of patient at the beginning of the study

obstruct (binary): obstruction of colon by tumor

outcome (binary): 1 if patient died within 5 years

TRTTMT (binary): if patient was treated

In [14]:
data.head()

Unnamed: 0  Sex  age  obstruct  outcome  TRTTMT
0           1    0   57         0        1    True
1           2    1   68         0        0   False
2           3    0   72         0        0    True
3           4    0   66         1        1    True
4           5    1   69         0        1   False

In [15]:
data.drop('Unnamed: 0', inplace=True, axis=1)

In [16]:
data.isna().sum()

Sex         0
age         0
obstruct    0
outcome     0
TRTTMT      0
dtype: int64

# Logistic Regression

In [17]:
y = data["outcome"]

X = data.drop('outcome', axis=1)

classifier = LogisticRegression(solver='lbfgs').fit(X, y)

# Odds

odds = p/(1-p)

p : probability of events (such as death)
1-p : probability that the event doesn't happen (non death)


odds = σ(θTx) / (1−σ(θTx))

σ(θTx(i)) = 1 / (1+e(−θTx))

odds = e(θTx(i))

θT : coefficients
x(i) : features

# coefficient for each feature 

In [18]:
thetas = classifier.coef_
thetas

array([[-0.21704833,  0.0460642 ,  0.37798496, -0.418984  ]])

In [19]:
for i in range(len(X.columns)):
    print("Feature {:<9s}: coefficient = {:<10f}".format(X.columns[i], thetas[0, i]))

Feature Sex      : coefficient = -0.217048 
Feature age      : coefficient = 0.046064  
Feature obstruct : coefficient = 0.377985  
Feature TRTTMT   : coefficient = -0.418984 


# Odds Ratio

OR for binary variables, it's defined as the odds when the variable is 1 divided by the odds when the variable is 0

ORxj = e(θj)

In [20]:
odds_ratios = np.exp(thetas)
odds_ratios

array([[0.80489107, 1.04714164, 1.45934099, 0.65771472]])

In [21]:
for i in range(len(X.columns)):
    print("Feature {:<10s}: coefficient = {:<10f} // OR = {:.2f}".format(X.columns[i], thetas[0, i], odds_ratios[0, i]))

Feature Sex       : coefficient = -0.217048  // OR = 0.80
Feature age       : coefficient = 0.046064   // OR = 1.05
Feature obstruct  : coefficient = 0.377985   // OR = 1.46
Feature TRTTMT    : coefficient = -0.418984  // OR = 0.66


Sex : OR = 0.80 < 1 donc être un homme réduit le risque de décès 

age : OR = 1.05 > 1 donc être jeune réduit le risque de décès

obstruct : OR = 1.46 > 1 donc ne pas présenter de tumeur obstructive du côlon réduit le risque de décès

TRTTMT : OR = 0.66 donc prendre un traitement réduit le risque de décès